### ISSUE TITLE
Prevent a single slow node from affecting entire cluster due to slow replica writes

### ISSUE THREAD

**Issue by vigyasharma** [elastic/elasticsearch#29673]
### Problem Description

Currently, when a bulk request arrives on a coordinating node, it is split into sub requests and sent into respective primary shards on different nodes. Once a write completes on primary shard, the primary shard node sends out requests to replica nodes and waits for them to complete.
If a primary shard is busy, i.e. its thread pool queue is full, the request is rejected with a 429 error. However, replica requests are sent with forced execution and infinite timeout. If the replica shard's thread pool queue is full, this request simply increases the queue's size and adds itself (unbounded queue). The timeout for such requests is also set to null, which implies no timeout. If the replica shard is on a slow / bad node, these requests would wait indefinitely.

While waiting for replica requests, threads and thread pool queues at coordinating and primary shard nodes are freed up, so they can accept more requests. However, the payload for these requests is still kept in memory, `to be resilient to failures of the node holding the primary`
In case the cluster has a single bad node where writes are not progressing due to disk issues, this setup consumes memory at other nodes of the cluster as well, eventually taking them down.
Meanwhile, client requests time out and the writes elasticsearch was waiting on, are presumed failed by clients.

Elasticsearch depends on async replica writes for maintaining consistency between primary and replica shards. It cannot timeout or abandon replica writes *once it has written to primary shard*, since that will diverge primaries and replicas.

Former discussion on this issue is listed here -- https://github.com/elastic/elasticsearch/issues/29531

### Proposed Optimization
The coordinating node will check both primary and replica queues before sending out a request, and **reject if either of the queues is full**; that is before even writing to the primary shard. Client will see a 429 response for operations rejected due to busy primary or replica shards. Thread pool queues will be checked via local copy of cluster state (``` _cat/thread_pool?local ```) to avoid an overload of network calls for each request.

Additionally, the coordinating node will retry sending this sub-request (by checking queues at all p and r shards) till a **configurable retry timeout**, to avoid aggressive request rejects. We will also make this **entire setting optional** and configurable dynamically via cluster settings, to let users decide whether to enable or disable it.

Existing implementation will continue after the coordinating node decides that queues have space and request is safe to process. It can still get rejected at primary shard node if thread pool gets full by the time the request arrives.

#### Pros
* This will prevent a single slow node from impacting other healthy nodes in the cluster.

#### Cons
* We will now throttle earlier than current implementation. It is possible that queues are currently full, but by the time a request actually requests a thread, it can be accommodated. However, this margin of error works both ways and should get cancelled out. It is also possible that the queue is seen empty at coordinating node, and gets full by the time the request actually arrives on the node.
* We plan to use local cluster state to prevent too many network calls. This means we have an inherent staleness in thread pool queues we refer, which can get aggravated during network partitions or slow networks. We will throttle more aggressively in such scenarios. However, slower than usual updates to cluster state are indicators of bigger cluster wide problems or overloaded clusters (master node), and throttling requests does not worsen the cluster's condition. Users can also chose to disable this setting if required.

Please let us know your thoughts on this. We would like to work on this and raise a pull request, if this seems in line with overall elasticsearch direction.

---

**Elasticsearch version** (`bin/elasticsearch --version`): <=5.5
**JVM version** (`java -version`): 1.8
**OS version** (`uname -a` if on a Unix-like system): Linux

**Comment by bleskes** [comment]
@vigyasharma thanks but can you please continue the discussion on the old issue? I don't see any fundamental difference that requires a new issue? I'm going to close it and respond on the other issue.
